liesel/model/model.py

```diff
@@ -235,12 +235,12 @@ def add(self, *args: Node | Var | GraphBuilder) -> GraphBuilder:
 
         for arg in args:
             if isinstance(arg, Node):
-                self.nodes.add(arg)
+                self.nodes.append(arg)
             elif isinstance(arg, Var):
-                self.vars.add(arg)
+                self.vars.append(arg)
             elif isinstance(arg, GraphBuilder):
-                self.nodes.update(arg.nodes)
-                self.vars.update(arg.vars)
+                self.nodes.extend(arg.nodes)
+                self.vars.extend(arg.vars)
             else:
                 raise RuntimeError(f"Cannot add {type(arg).__name__} to graph builder")
```

Comment:
The set has no duplicates but the list could hold the same node twice. Do we assume unique elements somewhere else in the code so that we should check for that here?

Reply:
Uniqueness is ensured [in the `_all_nodes_and_vars()` method](https://github.com/liesel-devs/liesel/blob/deterministic-distreg-mcmc/liesel/model/model.py#L176-L178), which is used for all relevant operations (at least, I hope so)...

Reply:
i'm not sure if the function does it. consider two different nodes which have the same input node, the input node will be twice in the node list.

see line 198f.
> nodes.extend(node.all_input_nodes())

Reply:
It does. We are filling the `all_nodes` list here, `nodes` is only the list we're iterating over:

``` python
while nodes:
    node = nodes.pop()

    if node in all_nodes:
        continue

    nodes.extend(node.all_input_nodes())
    all_nodes.append(node)
```

The relevant line is `if node in all_nodes: ...`.